edb/pgsql/compiler/stmt.py

```diff
@@ -49,8 +50,27 @@ def compile_SelectStmt(
 
         query = ctx.stmt
 
-        if not isinstance(stmt.result.expr, irast.MutatingStmt):
-            iterators = irutils.get_iterator_sets(stmt)
+        iterators = irutils.get_iterator_sets(stmt)
+        if iterators and irutils.contains_dml(stmt):
+            # If we have iterators and we contain nested DML
```

Comment:
We need to prove that the iterators are hoist-able, i.e. that they're not correlated with anything except among themselves.